src/include/txn.i

```diff
@@ -353,14 +354,17 @@ __wt_txn_id_alloc(WT_SESSION_IMPL *session, bool publish)
 
 	if (publish) {
 		session->txn.id = id;
-		WT_PUBLISH(WT_SESSION_TXN_STATE(session)->id, id);
+		WT_SESSION_TXN_STATE(session)->id = id;
 	}
 
-	while (txn_global->current != id ||
-	    !__wt_atomic_casv64(&txn_global->current, id, id + 1))
-		__wt_yield();
+	for (i = 0; txn_global->current != id; i++)
+		if (i < 100)
+			WT_PAUSE();
+		else
+			__wt_yield();
 
-	 return (id);
+	WT_PUBLISH(txn_global->current, id + 1);
```

Comment:
The reasoning for the change (apart from spinning) is that by the time we get here, we know `current == id`, so we don't need an atomic compare and swap.